x/incentives/keeper/distribute_test.go

```diff
@@ -177,92 +184,261 @@ func (suite *KeeperTestSuite) TestSyntheticDistribute() {
+			},
+			numEpochsPaidOver: 3,
+		},
+	}
 
-	// setup a non perpetual lock and gauge
-	_, gaugeID, gaugeCoins, startTime := suite.SetupLockAndGauge(false)
+	for _, tc := range tests {
+		suite.SetupTest()
 
-	// check that the sum of coins yet to be distributed is equal to the newly created gaugeCoins
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins)
+		incentivesKeeper := suite.App.IncentivesKeeper
 
-	// add coins to the previous gauge and check that the sum of coins yet to be distributed includes these new coins
-	addCoins := sdk.Coins{sdk.NewInt64Coin("stake", 200)}
-	suite.AddToGauge(addCoins, gaugeID)
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins.Add(addCoins...))
+		// check that the sum of coins yet to be distributed is nil
+		moduleToDistributeCoins := incentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
+		suite.Require().Equal(moduleToDistributeCoins, sdk.Coins(nil))
+		// gauges used in this test
+		gauges := []*types.Gauge{}
 
-	// create a new gauge
-	// check that the sum of coins yet to be distributed is equal to the gauge1 and gauge2 coins combined
-	_, _, gaugeCoins2, _ := suite.SetupNewGauge(false, sdk.Coins{sdk.NewInt64Coin("stake", 1000)})
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins.Add(addCoins...).Add(gaugeCoins2...))
+		expModuleToDistributeCoins := sdk.Coins{}
+		// coins that will be distributed after the first epoch for each gauge
+		expDistributedCoinsFirstEpoch := map[uint64]sdk.Coins{}
 
-	// move all created gauges from upcoming to active
-	suite.Ctx = suite.Ctx.WithBlockTime(startTime)
-	gauge, err := suite.App.IncentivesKeeper.GetGaugeByID(suite.Ctx, gaugeID)
-	suite.Require().NoError(err)
-	err = suite.App.IncentivesKeeper.MoveUpcomingGaugeToActiveGauge(suite.Ctx, *gauge)
-	suite.Require().NoError(err)
+		suite.LockTokens(lockOwner, sdk.Coins{sdk.NewInt64Coin("lptoken", 10)}, time.Second)
 
-	// distribute coins to stakers
-	distrCoins, err := suite.App.IncentivesKeeper.Distribute(suite.Ctx, []types.Gauge{*gauge})
-	suite.Require().NoError(err)
-	suite.Require().Equal(distrCoins, sdk.Coins{sdk.NewInt64Coin("stake", 105)})
+		// create new gauge using initialGaugeCoins
+		for _, coins := range tc.initialGaugeCoins {
+			_, gauge, gaugeCoins, _ := suite.SetupNewGauge(false, coins, tc.numEpochsPaidOver)
+			suite.Require().Equal(coins, gaugeCoins, "test %v", tc.name)
+
+			gauges = append(gauges, gauge)
+
+			moduleToDistributeCoins = incentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
+			expModuleToDistributeCoins = expModuleToDistributeCoins.Add(gaugeCoins...)
+			suite.Require().Equal(expModuleToDistributeCoins, moduleToDistributeCoins, "test %v", tc.name)
+
+			expDistributedCoinsFirstEpoch[gauge.Id] = coins
+		}
+
+		for gaugeID, coins := range tc.coinsToAddToGauges {
```

Comment:
Although `GetModuleToDistributeCoins` is consistently used here, I feel like there are several goals packages into a single test case: 
- test GetModuleToDistributeCoins
- test  gauge creation
- test  adding coins to gauge
- test  moving gauge from upcoming to active

Does it make sense to have a separate fixture with its own setup for each goal listed above?

Reply:
I'm good with either options.

Reply:
Could you help me understand why we need one large test? I'm totally okay with it as long as I understand the reasons behind this choice.

In my view of things, a unit test is called this way because it is meant to test a single unit of computation (a function or a method). Here, the test is called `TestGetModuleToDistributeCoins` implying that it tests `GetModuleToDistributeCoins`. However, there is a lot of other functionality being tested. As a result, I'm having a hard time evaluating what it is that is the goal of this test.

In particular, I don't understand why do we do both ` suite.SetupNewGauge()` and `suite.AddToGauge()` while testing `GetModuleToDistributeCoins`. If the goal is to test `GetModuleToDistributeCoins` only,  having `suite.SetupNewGauge()` is enough IMO.

Additionally, moving gauges to active and distributing feels like a completely separate test case from just testing  `GetModuleToDistributeCoins`.

It feels to me that each for loop in `TestGetModuleToDistributeCoins` deserves its own fixture like I described above:
- basic test for GetModuleToDistributeCoins
- test gauge creation
- test adding coins to gauge
- test moving gauge from upcoming to active and distributing

In my opinion, right now, a single large test is doing too many things.

Please let  me know what you think

Reply:
You've been doing fantastic with this PR @catShaark ! Thanks for all the thoroughness in applying and discussing suggestions